twilight-gateway/src/stream.rs

```diff
@@ -0,0 +1,504 @@
+    assert!(
+        concurrency < total,
+        "concurrency must be less than the total"
+    );
+
+    let concurrency = concurrency.try_into().unwrap();
+    let tls = TlsContainer::new().unwrap();
+
+    (bucket_id..total)
+        .step_by(concurrency)
+        .map(|index| {
+            let id = ShardId::new(index, total);
+            let mut config = per_shard_config(id);
+            config.set_tls(tls.clone());
+
+            Shard::with_config(id, config)
+        })
+        .collect::<FuturesUnordered<_>>()
+}
+
+/// Start a range of shards with provided configuration for each shard.
+///
+/// Lower end of the range must be less than the higher end.
+///
+/// Shards will all share the same TLS connector to reduce memory usage.
+///
+/// # Panics
+///
+/// Panics if the start is more than the end, the start is more than the total,
+/// or the end is more than the total.
+///
+/// Panics if loading TLS certificates fails.
+#[track_caller]
+pub fn start_range<F: Fn(ShardId) -> Config>(
+    range: impl RangeBounds<u64>,
+    total: u64,
+    per_shard_config: F,
+) -> impl Stream<Item = Result<Shard, ShardInitializeError>> + Send + 'static {
+    let range = calculate_range(range, total);
+    let tls = TlsContainer::new().unwrap();
+
+    range
+        .map(|index| {
+            let id = ShardId::new(index, total);
+            let mut config = per_shard_config(id);
+            config.set_tls(tls.clone());
+
+            Shard::with_config(id, config)
+        })
+        .collect::<FuturesUnordered<_>>()
+}
+
+/// Start all of the shards recommended for Discord in a single group.
+///
+/// Shards will all share the same TLS connector to reduce memory usage.
+///
+/// # Examples
+///
+/// Start all of the shards recommended by Discord and collect them into a map:
+///
+/// ```no_run
+/// use futures::StreamExt;
+/// use std::{collections::HashMap, env, future};
+/// use twilight_gateway::{stream, Config, Intents};
+///
+/// # #[tokio::main]
+/// # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// let token = env::var("DISCORD_TOKEN")?;
+///
+/// // callback to create a config for each shard, useful for when not all shards
+/// // have the same configuration, such as for per-shard presences
+/// let config_callback = |_| Config::new(token.clone(), Intents::GUILDS);
+///
+/// let shards = stream::start_recommended(token.clone(), config_callback)
+///     .await?
+///     .filter_map(|shard_result| async move {
+///         shard_result.ok().map(|shard| (shard.id().number(), shard))
+///     })
+///     .collect::<HashMap<_, _>>()
+///     .await;
+///
+/// println!("total shards: {}", shards.len());
+/// # Ok(()) }
+/// ```
+///
+/// # Errors
+///
+/// Returns a [`StartRecommendedErrorType::Deserializing`] error type if the
+/// response body failed to deserialize.
+///
+/// Returns a [`StartRecommendedErrorType::Request`] error type if the request
+/// failed to complete.
+#[track_caller]
```

Comment:
```suggestion
/// Returns a [`StartRecommendedErrorType::Request`] error type if the request
/// failed to complete.
///
/// # Panics
///
/// Panics if loading TLS certificates fails.
#[track_caller]
```

I don't think we should add the other transitively panic since it should not happen when using this function, but I will leave that up to you

Reply:
It should not happen is good enough for me.